packages/beacon-state-transition/src/altair/block/processAttestation.ts

```diff
@@ -17,69 +16,80 @@ import {
-  }
+    validateAttestation(state as CachedBeaconState<allForks.BeaconState>, attestation);
 
-  // this check is done last because its the most expensive (if signature verification is toggled on)
-  if (
-    !isValidIndexedAttestation(
-      state as CachedBeaconState<allForks.BeaconState>,
-      epochCtx.getIndexedAttestation(attestation),
-      verifySignature
-    )
-  ) {
-    throw new Error("Attestation is not valid");
-  }
-  const {timelySource, timelyTarget, timelyHead} = getAttestationParticipationStatus(
-    state,
-    data,
-    state.slot - data.slot
-  );
+    // Retrieve the validator indices from the attestation participation bitfield
+    const attestingIndices = epochCtx.getAttestingIndices(data, attestation.aggregationBits);
 
-  // Retrieve the validator indices from the attestation participation bitfield
-  const attestingIndices = epochCtx.getAttestingIndices(data, attestation.aggregationBits);
+    // this check is done last because its the most expensive (if signature verification is toggled on)
+    // TODO: Why should we verify an indexed attestation that we just created? If it's just for the signature
+    // we can verify only that and nothing else.
+    if (verifySignature) {
+      const sigSet = getAttestationWithIndicesSignatureSet(
+        state as CachedBeaconState<allForks.BeaconState>,
+        attestation,
+        attestingIndices
+      );
+      if (!verifySignatureSet(sigSet)) {
+        throw new Error("Attestation signature is not valid");
+      }
+    }
 
-  // For each participant, update their participation
-  // In epoch processing, this participation info is used to calculate balance updates
-  let totalBalancesWithWeight = BigInt(0);
-  for (const index of attestingIndices) {
-    const status = epochParticipation.getStatus(index) as IParticipationStatus;
-    const newStatus = {
-      timelySource: status.timelySource || timelySource,
-      timelyTarget: status.timelyTarget || timelyTarget,
-      timelyHead: status.timelyHead || timelyHead,
-    };
-    epochParticipation.setStatus(index, newStatus);
-    /**
-     * Spec:
-     * baseReward = state.validators[index].effectiveBalance / EFFECTIVE_BALANCE_INCREMENT * baseRewardPerIncrement;
-     * proposerRewardNumerator += baseReward * totalWeight
-     */
-    const totalWeight =
-      BigInt(!status.timelySource && timelySource) * TIMELY_SOURCE_WEIGHT +
-      BigInt(!status.timelyTarget && timelyTarget) * TIMELY_TARGET_WEIGHT +
-      BigInt(!status.timelyHead && timelyHead) * TIMELY_HEAD_WEIGHT;
-    if (totalWeight > 0) {
-      totalBalancesWithWeight += state.validators[index].effectiveBalance * totalWeight;
+    const epochParticipation =
+      data.target.epoch === epochCtx.currentShuffling.epoch
+        ? state.currentEpochParticipation
+        : state.previousEpochParticipation;
+
+    const {timelySource, timelyTarget, timelyHead} = getAttestationParticipationStatus(
```

Comment:
attestations in same block tend to have same source/target/head, could you find a way to cache block roots queried from the state? it's not efficient to keep calling `state.blockRoots[i]` multiple times for same value

Reply:
Agree, but we should tackle that in another PR, not here